File: python/20_valid-parentheses.py
from collections import deque
class Solution:
    def isValid(self, s: str) -> bool:
        # [] {} ()
        dq=deque()
        for cha in s:
            if cha in ('[','{','('):
                dq.append(cha)
            else:
                if not dq:
                    return False
                pre=dq.pop()
                if cha==']' and pre!='[':
                    return False
                if cha=='}' and pre!='{':
                    return False
                if cha==')' and pre!='(':
                    return False
        return True if not dq else False

File: go/hot_100/20_valid-parentheses.go
package main

func isValid(s string) bool {
	slice := []string{}
	for _, v := range s {
		if string(v) == "{" || string(v) == "[" || string(v) == "(" {
			slice = append(slice, string(v))
		} else {
			if len(slice) == 0 {
				return false
			}
			pre := slice[len(slice)-1]
			if string(v) == ")" && pre != "(" || string(v) == "]" && pre != "[" || string(v) == "}" && pre != "{" {
				return false
			}
			slice = slice[:len(slice)-1]
		}
	}
	return len(slice) == 0
}
